<?xml version="1.0"?>
<robot name="rover_lidar" xmlns:xacro="http://ros.org/wiki/xacro">
    <!-- Define robot constants -->
    <xacro:property name="base_width" value="0.28"/>
    <xacro:property name="base_length" value="0.58"/>
    <xacro:property name="base_height" value="0.08"/>
    <xacro:property name="tower_side" value="0.06"/>
    <xacro:property name="tower_height" value="0.24"/>

    <xacro:property name="wheel_radius" value="0.0593"/>
    <xacro:property name="wheel_width" value="0.056"/>
    <xacro:property name="wheel_ygap" value="0.032"/>
    <xacro:property name="wheel_zoff" value="0.087"/>
    <xacro:property name="wheel_xoff" value="0.0"/>

    <xacro:property name="imu_x" value="-0.06"/>
    <xacro:property name="imu_y" value="-0.06"/>
    <xacro:property name="imu_z" value="0.20"/>


    <!-- Robot Base -->
    <link name="base_link">
        <visual>
        <geometry>
            <box size="${base_length} ${base_width} ${base_height}"/>
        </geometry>
        <material name="Orange">
            <color rgba="1.0 1.0 0.0 1.0"/>
        </material>
        </visual>
    </link>

    <!-- Robot Footprint -->
    <link name="base_footprint"/>

    <joint name="base_joint" type="fixed">
        <parent link="base_footprint"/>
        <child link="base_link"/>
        <origin xyz="0.0 0.0 ${(wheel_radius+wheel_zoff)}" rpy="0 0 0"/>
    </joint>

    <!-- Tower -->
    <link name="tower_link">
        <visual>
        <geometry>
            <box size="${tower_side} ${tower_side} ${tower_height}"/>
        </geometry>
        <material name="Orange">
            <color rgba="1.0 1.0 0.0 1.0"/>
        </material>
        </visual>
    </link>

    <joint name="tower_joint" type="fixed">
        <parent link="base_link"/>
        <child link="tower_link"/>
        <origin xyz="0.0 0.0 ${tower_height/2+base_height/2}" rpy="0 0 0"/>
    </joint>

    <!-- IMU -->
    <link name="imu_link">
        <visual>
        <geometry>
            <box size="${0.05} ${0.02} ${0.01}"/>
        </geometry>
        <material name="Cyan">
            <color rgba="0.0 1.0 1.0 1.0"/>
        </material>
        </visual>
    </link>

    <joint name="imu_joint" type="fixed">
        <parent link="base_link"/>
        <child link="imu_link"/>
        <origin xyz="${imu_x} ${imu_y} ${imu_z}" rpy="0 0 0"/>
    </joint>

    <!-- LIDAR -->
    <link name="lidar_link">
        <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
            <cylinder radius="0.05" length="0.1"/>
        </geometry>
        <material name="Cyan">
            <color rgba="0.0 1.0 1.0 1.0"/>
        </material>
        </visual>
    </link>

    <joint name="lidar_joint" type="fixed">
      <parent link="tower_link"/>
      <child link="lidar_link"/>
      <origin xyz="0 0 ${tower_height/2+0.05}" rpy="0 0 ${pi/2}"/>
    </joint>

    <!-- Wheels -->
    <xacro:macro name="wheel" params="prefix x_reflect y_reflect">
        <link name="${prefix}_link">
            <visual>
                <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
                <geometry>
                    <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
                </geometry>
                <material name="Gray">
                <color rgba="0.5 0.5 0.5 1.0"/>
                </material>
            </visual>
        </link>

        <joint name="${prefix}_joint" type="continuous">
            <parent link="base_link"/>
            <child link="${prefix}_link"/>
            <origin xyz="${x_reflect*wheel_xoff} ${y_reflect*(base_width/2+wheel_ygap)} ${-wheel_zoff}" rpy="0 0 0"/>
            <axis xyz="0 1 0"/>
        </joint>
    </xacro:macro>

    <xacro:wheel prefix="drivewhl_l" x_reflect="-1" y_reflect="1" />
    <xacro:wheel prefix="drivewhl_r" x_reflect="-1" y_reflect="-1" />
</robot>